iothub/device/src/InternalClient.cs

```diff
@@ -1895,7 +1895,11 @@ public void Dispose()
             _fileUploadHttpTransportHandler?.Dispose();
             _deviceReceiveMessageSemaphore?.Dispose();
             _twinDesiredPropertySemaphore?.Dispose();
-            IotHubConnectionString?.TokenRefresher?.Dispose();
+
+            if ((IotHubConnectionString?.TokenRefresher?.ShouldSdkDisposeInstance).GetValueOrDefault())
+            {
+                IotHubConnectionString?.TokenRefresher?.Dispose();
```

Comment:
My point yesterday during chat was that, this logic could be centralized inside the `TokenRefresher.Dispose()` call, ensuring that logic doesn't need to be be right anywhere else.

Reply:
Do you mean that the check for "should this be disposed by sdk" be managed inside the `TokenRefresher.Dispose()` call? How would that work though? 
For scenarios where we don't initialize the instance, "should this be disposed by sdk" would resolve to false and we wouldn't release the managed resources. But then, how would the user make the call to release those resources? They would still call `TokenRefresher.Dispose()`, but the expression there would still evaluate to false.

Since the intended behavior on dispose is the same, wouldn't it make sense for the sdk to just evaluate if it should be calling dispose, and call it if required?

Reply:
Good call. It would have to know who is making the call, which would require some tricks.